airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.testutils;
+
+import io.airbyte.commons.logging.LoggingHelper;
+import io.airbyte.commons.logging.MdcScope;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.JdbcDatabaseContainer;
+import org.testcontainers.containers.output.Slf4jLogConsumer;
+import org.testcontainers.utility.DockerImageName;
+
+/**
+ * This class wraps a specific shared testcontainer instance, which is created exactly once.
+ */
+class ContainerFactoryWrapper<C extends JdbcDatabaseContainer<?>> {
+
+  static private final Logger LOGGER = LoggerFactory.getLogger(ContainerFactoryWrapper.class);
+  static private final ConcurrentHashMap<String, ContainerFactoryWrapper<?>> SHARED_SINGLETONS = new ConcurrentHashMap<>();
+
+  static public final MdcScope.Builder TESTCONTAINER_LOG_MDC_BUILDER = new MdcScope.Builder()
+      .setLogPrefix("testcontainer")
+      .setPrefixColor(LoggingHelper.Color.RED_BACKGROUND);
+
+  @SuppressWarnings("unchecked")
+  static <C extends JdbcDatabaseContainer<?>> C getOrCreateShared(ContainerFactory<C> factory, String imageName, String... methods) {
```

Comment:
That's super weird to have a different C here than in the class definition.

Reply:
It's static

Reply:
I feel like all the methods here should be static, and the main class shouldn't be parameterized, no?

Reply:
You're right, it's better when the class isn't parameterized.